commit 8b5ee2c66d5c4c1312fd193d4138e6963160ba43
Author: Tiezhu Yang <yangtiezhu@loongson.cn>
Date:   Mon May 1 17:19:52 2023 +0800

    LoongArch: Add support for function error injection
    
    Inspired by the commit 42d038c4fb00f ("arm64: Add support for function
    error injection") and the commit ee55ff803b383 ("riscv: Add support for
    function error injection"), this patch supports function error injection
    for LoongArch.
    
    Mainly implement two functions:
    (1) regs_set_return_value() which is used to overwrite the return value,
    (2) override_function_with_return() which is used to override the probed
    function returning and jump to its caller.
    
    Here is a simple test under CONFIG_FUNCTION_ERROR_INJECTION and
    CONFIG_FAIL_FUNCTION:
    
      # echo sys_clone > /sys/kernel/debug/fail_function/inject
      # echo 100 > /sys/kernel/debug/fail_function/probability
      # dmesg
      bash: fork: Invalid argument
      # dmesg
      ...
      FAULT_INJECTION: forcing a failure.
      name fail_function, interval 1, probability 100, space 0, times 1
      ...
      Call Trace:
      [<90000000002238f4>] show_stack+0x5c/0x180
      [<90000000012e384c>] dump_stack_lvl+0x60/0x88
      [<9000000000b1879c>] should_fail_ex+0x1b0/0x1f4
      [<900000000032ead4>] fei_kprobe_handler+0x28/0x6c
      [<9000000000230970>] kprobe_breakpoint_handler+0xf0/0x118
      [<90000000012e3e60>] do_bp+0x2c4/0x358
      [<9000000002241924>] exception_handlers+0x1924/0x10000
      [<900000000023b7d0>] sys_clone+0x0/0x4
      [<90000000012e4744>] do_syscall+0x7c/0x94
      [<9000000000221e44>] handle_syscall+0xc4/0x160
    
    Tested-by: Hengqi Chen <hengqi.chen@gmail.com>
    Acked-by: Masami Hiramatsu (Google) <mhiramat@kernel.org>
    Signed-off-by: Tiezhu Yang <yangtiezhu@loongson.cn>
    Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>

diff --git a/arch/loongarch/Kconfig b/arch/loongarch/Kconfig
index 3e5d6acbf240..629a6dac33e1 100644
--- a/arch/loongarch/Kconfig
+++ b/arch/loongarch/Kconfig
@@ -100,6 +100,7 @@ config LOONGARCH
 	select HAVE_FAST_GUP
 	select HAVE_FTRACE_MCOUNT_RECORD
 	select HAVE_FUNCTION_ARG_ACCESS_API
+	select HAVE_FUNCTION_ERROR_INJECTION
 	select HAVE_FUNCTION_GRAPH_TRACER
 	select HAVE_FUNCTION_TRACER
 	select HAVE_GENERIC_VDSO